System overview documentation (#26)
FS-Online/Driverless-Competition-Simulator: docs/system-overview.md

_This document is intended to read top-to-bottom. Do yourself a favour and read the whole thing without skipping ;)_

# Formula Student Driverless Simulation: System overview

FSDS is built around Unreal Engine 4 (the game engine) and the AirSim plugin. 
The game engine does all the graphical rendering, collision simulation and car movement simulation. 
A separate component - the simulator - will handle all controll the simulation, provide external interfaces and store what is happening.

! There will be also something that provides a video feed for livestreaming the simulation. 
  Since it is still unclear how this will work this is not included in this description.

![System overview](images/system-overview.png)

## Autonomous Systems.

Every autonomous system run on their own seperated environment. 
Idealy this would be separate virtual machines but also Docker containers could be used.

> The autonomous systems are expected to continuously run a ROS master.

Reviewer: Has this been decided?

Author: The alternative would be that the autonomous systems are only running a ROS master when they are 'in use'. This would result in having the simulator somehow start the ros autonomous system within the team's environment. Since ROS is not running at that point, we would have to setup another communication channel to signal that system to start it's ros core. An extra communication channel is a lot of work to build, test and document for both us and the teams. That's why I think we should definitely go for having the autonomous systems run their ros masters continuously. How the simulator will deal with this is indeed an open question but I believe something we can figure out.

Reviewer: let's start with this concept in our testing environment then!

When the simulator is ready to do a drive for the given autonomous system, the simulator will launch a ROS node connected to this autonomous system's ROS master.
This ROS node will exist outside of the autonomous system's environment, instead it will run on the simulator computer.
The simulator ROS node will publish sensor data and listen for car setpoints on a set of topics defined [here](airsim-ros-wrapper.md).
When the simulation is finished, the ROS node will disconnect from the autonomous system's ROS master and sensor data is no longer streaming.

During the competition the teams will not be allowed to access their autonomous systems.
All remote access to the environments will be cut off completely.

> So every autonomous system must be able to run multiple missions without human interferance.

Reviewer 2: ```suggestion
So every autonomous system must be able to run multiple missions without human interference.
```

To let the simulation know what is expected from it, the simulator will send mission control messages a few seconds before the event start.
These messages contain information about the mission (trackdrive, autocross, etc).
When the autonomous system receives a mission message it can expect to receive sensor data shortly after.

> Since the laptimer for all events start whenever the car crosses the start-line, the autonomous system's can take all the time they need to launch their relevant algorithms for the mission (within reasonable time).

Reviewer 2: ```suggestion
Since the laptimer for all events starts whenever the car crosses the start-line, the autonomous systems can take all the time they need to launch their relevant algorithms for the mission (within reasonable time).
```

## The Simulator

The simulator is a continuously running program that is like the spider in the web.

The simulator it offers a webinterface to the event's officials. 
Using this webinterface the officials can choose what team/car is going to drive and which track is going to be used.
The official will also be able to send the start signal, view lap times, down or out cones, car off course's.
There is also an emergency stop button in case the car is uncontrollable.

The simulator connects to the simulated world inside Unreal Engine using AirSim (more on that later).
It also connects to the autonomous systems and acts like a bridge between the two.

> On one hand it requests sensordata and passes it along on ros topics to the current autonomous system.

Reviewer 3: via ros topics

actually I'm not 100% sure if its "via" or "on"

> On the other hand it receives car controll commands from the autonomous system and forwards it to the virtual world.

Reviewer 3: controll

> Only 1 car will be able to run on this autonomous system.

Reviewer 3: Only 1 car at a time

The wording makes it sound like only 1 car will race for the whole competition ;)

> The official can select what team, and thus what autonomous system, is currently selected.

Reviewer 3: which team which autonomous system

> When the team changes not only the autonomous system, but also the car inside the virtual world changes.

Reviewer 3: inside the virtual world will change.

> When the team changes not only the autonomous system, but also the car inside the virtual world changes.

Reviewer 2: ```suggestion
When the team changes, not only does this change the autonomous system, but also the car inside the virtual world.
```

The sensor suite (sensor types and locations) and the car livery (looks of the car) are updated.
The simulator keeps track of these details and passes them along to the virtual world to ensure acurate representation.

When a start signal is received the simulator informs the current autonomous system of the mission and starts sending sensordata it's way.
When a stop signal is received, the simulator stops interacting with the attached autonomous system and informs the virtual world that the car must stop driving

During a mission the simulator keeps polling the world for 'referee state'. 
This is information that in a phisical world would be relevant to the referee.
Currently this includes a list of down or out cones and the timestamp of when they went down or out, a list of laptimes and a list of when the car went off-track.
More information will for sure be added.

> In a single logbook is stored what happend inside the simulation.

Reviewer 3: Change the order: What happens inside the simulation is stored in a single logbook.

This includes all referee updates, what autonomous systems were selected and what tracks were used.

> If something unexpectedly went wrong like a system's crash or error, a short description of what happend is included in this logbook.

Reviewer 3: unexpected

> It gives a timeline of everything that happend to always go back afterwards and check what happend.

Reviewer 3: It gives a timeline of everything that happend, so that one can always check it afterwards.

Something about this sentence feels off, but I'm not sure how to rephrase it.

> The logbook is stored on disk so that even when the whole system crashes we always have the logbook.

Reviewer 3: so that in the event of a whole system crash we will still have the logbook.

On that note, make sure to flush the logbook data often.

It is also shown within the referee's web interface.

## The virtual world (Unreal Engine and AirSim)

The actual simulation takes place inside an Unreal Engine 4 world.
Unreal takes care of the heavy lifting involved with a real-life simulation.

> All phisics, lighting and world building are handled by Unreal.

Reviewer 2: ```suggestion
All physics, lighting and world building are handled by Unreal.
```

AirSim is used to connect Unreal with the Simulator.
This plugin is added inside the Unreal world and takes controll of most of the game logic.
It receives the sensor suite and simulates the sensors, it moves the car according to trajectory setpoints and exposes an api for external management.

> This rpc api is used by the simulator to interact with the world.

Reviewer 3: You dont mention rpc in the previous sentance.
